android/src/main/java/com/bugsnag/JavaScriptException.java

```diff
@@ -30,47 +30,108 @@ public void toStream(JsonStream writer) throws IOException {
+        writer.endArray();
+        writer.endObject();
+    }
+
+    private void serialiseJsCoreFrame(JsonStream writer, String frame) throws IOException {
+        // expected format is as follows:
+        //   release:
+        //     "$method@$filename:$lineNumber:$columnNumber"
+        //   dev:
+        //     "$method@?uri:$lineNumber:$columnNumber"
+
+        writer.beginObject();
+        String[] methodComponents = frame.split("@", 2);
+        String fragment = methodComponents[0];
+        if (methodComponents.length == 2) {
+            writer.name("method").value(methodComponents[0]);
+            fragment = methodComponents[1];
+        }
+
+        int columnIndex = fragment.lastIndexOf(":");
+        if (columnIndex != -1) {
+            String columnString = fragment.substring(columnIndex + 1);
+            Integer columnNumber = parseIntSafe(columnString);
+
+            if (columnNumber != null) {
+                writer.name("columnNumber").value(columnNumber);
             }
+            fragment = fragment.substring(0, columnIndex);
+        }
+
+        int lineNumberIndex = fragment.lastIndexOf(":");
+        if (lineNumberIndex != -1) {
+            String lineNumberString = fragment.substring(lineNumberIndex + 1);
+            Integer lineNumber = parseIntSafe(lineNumberString);
+
+            if (lineNumber != null) {
+                writer.name("lineNumber").value(lineNumber);
+            }
+            fragment = fragment.substring(0, lineNumberIndex);
+        }
 
-            int lineNumberIndex = fragment.lastIndexOf(":");
-            if (lineNumberIndex != -1) {
-                String lineNumberString = fragment.substring(lineNumberIndex + 1);
-                try {
-                    int lineNumber = Integer.parseInt(lineNumberString);
+        writer.name("file").value(fragment);
+        writer.endObject();
+    }
+
+    private void serialiseHermesFrame(JsonStream writer, String frame) throws IOException {
+        // expected format is as follows:
+        //   release
+        //     "at $method (address at $filename:$lineNumber:$columnNumber)"
+        //   dev
+        //     "at $method ($filename:$lineNumber:$columnNumber)"
+
+        int srcInfoStart = Math.max(frame.lastIndexOf(" "), frame.lastIndexOf("("));
+        int srcInfoEnd = frame.lastIndexOf(")");
+        boolean hasSrcInfo = srcInfoStart > -1 && srcInfoStart < srcInfoEnd;
+
+        int methodStart = "at ".length();
+        int methodEnd = frame.indexOf(" (");
+        boolean hasMethodInfo = methodStart < methodEnd;
+
+        // serialise srcInfo
+        if (hasSrcInfo || hasMethodInfo) {
+            writer.beginObject();
+            writer.name("method").value(frame.substring(methodStart, methodEnd));
+            if (hasSrcInfo) {
+                String srcInfo = frame.substring(srcInfoStart + 1, srcInfoEnd);
+                String file = srcInfo.replaceFirst(":\\d+:\\d+$", "");
```

Comment:
There are a couple of regexes in this class that are hardcoded as strings - it'd be good to extract them to a `private static final` constant and possibly add a quick comment summarising what they do

Reply:
I would agree if the regexes were used more than once, but to move them away from the place they're used would be less clear in my opinion. I'm happy to name and comment them, but I'd be against moving the definitions away from the location they're used, if that makes sense?

Is this recommendation purely a stylistic preference or are you suggesting the performance would be improved by making it static too?

Reply:
Adding comments would resolve this IMO - as they are only used in one place a constant might be overkill.